webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/apidata/JsonQueryObjects.scala

```diff
@@ -0,0 +1,602 @@
+import com.normation.rudder.domain.queries.NodeReturnType
+import com.normation.rudder.domain.queries.QueryReturnType
+import com.normation.rudder.domain.queries.QueryTrait
+import com.normation.rudder.services.queries.CmdbQueryParser
+import com.normation.rudder.services.queries.JsonQueryLexer
+import com.normation.rudder.services.queries.StringCriterionLine
+import com.normation.rudder.services.queries.StringQuery
+import io.scalaland.chimney.dsl._
+
+/*
+ * This class deals with everything serialisation related for API.
+ * Change things with care! Everything must be versionned!
+ * Even changing a field name can lead to an API incompatible change and
+ * so will need a new API version number (and be sure that old behavior is kept
+ * for previous versions).
+ */
+
+
+///////////////////////////// decoder ///////////////
+
+/*
+ * Object used in JSON query POST/PUT request.
+ * For disambiguation, objects are prefixed by JQ.
+ *
+ * Note about id/revision: it does not make sense to try to change a
+ * revision, it's immutable. Only head can be changed, so rev is never
+ * specified in JQ object.
+ * It can make sense to clone a specific revision, so sourceId can have a
+ * sourceRevision too.
+ */
+object JsonQueryObjects {
+  import JsonResponseObjects.JRRuleTarget
+
+  final case class JQRuleCategory(
+        name       : Option[String] = None
+      , description: Option[String] = None
+      , parent     : Option[String] = None
+      , id         : Option[String] = None
+    ) {
+
+    def update(ruleCategory: RuleCategory) = {
+      ruleCategory.using(this).ignoreRedundantPatcherFields.patch
+    }
+  }
+
+
+  final case class JQDirectiveSectionVar(
+      name : String
+    , value: String
+  )
+  final case class JQDirectiveSection(
+      name    : String
+      // Map is necessary b/c of the "var" in {"vars": [ {"var": { "name":..., "value":... }}}
+    , sections: Option[List[Map[String, JQDirectiveSection]]]
+      // Map is necessary b/c of the "var" in {"vars": [ {"var": { "name":..., "value":... }}}
+    , vars    : Option[List[Map[String, JQDirectiveSectionVar]]]
+  ) {
+    def toSectionVal: (String, SectionVal) = {
+      ( name
+      , SectionVal(
+            sections.map(_.flatMap(_.get("section").map(s => s.toSectionVal)).groupMap(_._1)(_._2)).getOrElse(Map())
+          , vars.map(_.flatMap(_.get("var").map(x => (x.name, x.value))).toMap).getOrElse(Map())
+        )
+      )
+    }
+  }
+
+  // this one is a direction mapping of Directive with optionnal fields for patching
+  case class PatchDirective(
+      id              : Option[DirectiveId]
+    , techniqueVersion: Option[TechniqueVersion]
+    , parameters      : Option[Map[String, Seq[String]]]
+    , name            : Option[String]
+    , shortDescription: Option[String]
+    , policyMode      : Option[Option[PolicyMode]]
+    , longDescription : Option[String]
+    , priority        : Option[Int]
+    , _isEnabled      : Option[Boolean]
```

Comment:
why the _ ?

Reply:
it's the name of the field in the target object, the patch lib look-up field by name by default

Reply:
:+1: